src/main/java/com/pingcap/tikv/meta/TiSelectRequest.java

```diff
@@ -45,39 +43,72 @@
     private TiExpr having;
     private boolean distinct;
 
-      public SelectRequest toProto() {
-          return build();
-      }
-
-      public SelectRequest build() {
-        // add optimize later
-        // Optimize merge groupBy
-        this.fields.forEach(expr -> this.builder.addFields(expr.toProto()));
-        if(!this.fields.isEmpty()) {
-            // convert fields type to set for later usage.
-            Set<String> fieldSet = ImmutableSet.copyOf(TiFluentIterable.from(this.fields).transform(
-                    expr -> ((TiColumnRef)expr).getName()
-            ));
-            // solve
-            List<TiColumnInfo> colToRemove = new ArrayList<>();
-            this.tableInfo.getColumns().forEach(
-                            col -> {
-                                // remove column from table if such column is not in fields
-                                if(!fieldSet.contains(col.getName())) {
-                                    colToRemove.add(col);
-                                }
-                            }
-            );
+    public TiSelectRequest(SelectRequest.Builder builder) {
+        this.builder = builder;
+    }
 
-            this.tableInfo.getColumns().removeAll(colToRemove);
-        }
+    public SelectRequest build() {
+        List<TiColumnInfo> colToAdd = new ArrayList<>();
+        this.fields.forEach(expr -> {
+            this.builder.addFields(expr.toProto());
+            colToAdd.addAll(getColumnInfoFromExpr(expr));
+        });
         this.groupBys.forEach(expr -> this.builder.addGroupBy(expr.toProto()));
         this.orderBys.forEach(expr -> this.builder.addOrderBy(expr.toProto()));
-        this.aggregates.forEach(expr -> this.builder.addAggregates(expr.toProto()));
+        this.aggregates.forEach(expr -> {
+            this.builder.addAggregates(expr.toProto());
+            colToAdd.addAll(getColumnInfoFromExpr(expr));
+        });
         this.builder.setFlags(flags);
-        this.builder.setTableInfo(tableInfo.toProto());
+
+        TableInfo table = TableInfo.newBuilder()
+                .setTableId(tableInfo.getId())
+                .addAllColumns(colToAdd.stream().map(TiColumnInfo::toProto).collect(Collectors.toList()))
+                .build();
+        this.builder.setTableInfo(table);
         this.builder.setTimeZoneOffset(timeZoneOffset);
         this.builder.setStartTs(startTs);
         return this.builder.build();
+    }
+
+    private List<TiColumnInfo> getColumnInfoFromExpr(TiExpr expr) {
```

Comment:
@ilovesoup It would be better to do it later since you already create a `MetaUtil.java`  in your PR. To avoid merge conflict, it would be nice do it later.